### PR TITLE
feat (ui/solid): useAssistant, useObject, fix deep object updating

### LINKED ISSUE #3797: Tool results no longer being streamed to client in v4
### Description

1. Set up a streaming AI agent using SolidStart
2. Have the agent invoke a tool.
3. Observe as no tool results are streamed in

### Code example

My repository is here: https://github.com/spirit-led-software/theaistudybible

See:
1. apps/www/src/server/api/secure/chat.ts for the api
2. apps/www/src/hooks/use-chat.ts for the client

### AI provider

@ai-sdk/solid 1.0.1

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lgrammel**
## Summary

- add `useObject` for Solid.js
- add `useAssistant` for Solid.js
- fix `useChat` for Solid.js (deep object updates)

Continues #3823 & fixes #3797 

## Tasks

- [x] align examples
- [x] remove exposing messages directly (avoid having idiosyncratic APIs or breaking changes)
- [x] manual testing
- [x] review code
- [x] backports
- [x] docs updates
- [x] changeset

### CODE REVIEW

**Comment by lgrammel** [comment]
@ian-pascoe reverted a few changes around the api to prevent breakages and to keep the apis consistent between frameworks. please let me know if this addresses your issues or if i've reverted anything important that has unwanted side effects.

File: packages/solid/tsconfig.json [hunk 1]
```diff
@@ -1,8 +1,13 @@
...
   "compilerOptions": {
-    "target": "ES2018",
-    "stripInternal": true
+    "jsx": "preserve",
+    "jsxImportSource": "solid-js",
+    "lib": ["DOM", "DOM.Iterable", "ESNext"],
+    "target": "ESNext",
+    "module": "ESNext",
+    "stripInternal": true,
+    "moduleResolution": "Bundler"
```
Comment (lgrammel) [inline packages/solid/tsconfig.json:10]: @ian-pascoe why were these changes needed?

File: packages/ui-utils/src/process-chat-response.ts [hunk 2]
```diff
@@ -71,7 +71,7 @@ export async function processChatResponse({
     // trigger update for streaming by copying adding a revision id that changes
     // (without it, the changes get stuck in SWR and are not forwarded to rendering):
     const copiedMessage = {
-      ...currentMessage,
+      ...structuredClone(currentMessage), // deep copy
```
Comment (lgrammel) [inline packages/ui-utils/src/process-chat-response.ts:74]: @ian-pascoe is this key to your fix or optional?

File: packages/solid/tsconfig.json (hunk 1 above)
Comment (ian-pascoe) [reply packages/solid/tsconfig.json:10]: Package resolution for SSR types was off, these are the recommended settings for SolidJS

File: packages/ui-utils/src/process-chat-response.ts (hunk 2 above)
Comment (ian-pascoe) [reply packages/ui-utils/src/process-chat-response.ts:74]: This is actually the most important part of the fix

**Comment by ian-pascoe** [comment]
> @ian-pascoe reverted a few changes around the api to prevent breakages and to keep the apis consistent between frameworks. please let me know if this addresses your issues or if i've reverted anything important that has unwanted side effects.

https://github.com/vercel/ai/pull/3829/commits/b549bf2d786ab88c698e1ab737c57c6aaea4a859

These reverts are fine, but the reason I updated it is because when you use the accessor method, the options become reactive. With the current version, they are not.

Figured it was good to show the ability for reactive props in the examples

**Comment by lgrammel** [comment]
@ian-pascoe fair point. I'll leave them as is, because I want to keep the examples between the frameworks as aligned as possible and this seems like additional SolidJS functionality.

File: packages/ui-utils/src/process-chat-response.ts (hunk 2 above)
Comment (lgrammel) [reply packages/ui-utils/src/process-chat-response.ts:74]: wait a sec - we are already shallow cloning, wouldn't that help with referential integrity? this is why `revisionId` exists (for SWR).

Comment (ian-pascoe) [reply packages/ui-utils/src/process-chat-response.ts:74]: If you remove the deep clone and use the solidjs tool example or the solidjs annotations example, you'll see that those values do not get updates

**Review by jeremyphilemon — APPROVED** [review]

**Comment by tobiasstrebitzer** [comment]
This PR causes an (i believe) unintended side-effect, changing the `createdAt` type from `Date` to `string`.

**Implications:**
* Might cause runtime errors for users, e.g. `message.createdAt.getTime()`
* With typings still bound to `Date`, there won't be any eslint/ts warnings.

**Recommentation:**

```ts
const copiedMessage = {
  ...JSON.parse(JSON.stringify(currentMessage)),
  createdAt: currentMessage.createdAt, // <- NEW, to retain correct type of `createdAt`
  revisionId: generateId2()
};
```

**Comment by ian-pascoe** [comment]
> This PR causes an (i believe) unintended side-effect, changing the `createdAt` type from `Date` to `string`.
> 
> **Implications:**
> 
> * Might cause runtime errors for users, e.g. `message.createdAt.getTime()`
> * With typings still bound to `Date`, there won't be any eslint/ts warnings.
> 
> **Recommentation:**
> 
> ```ts
> const copiedMessage = {
>   ...JSON.parse(JSON.stringify(currentMessage)),
>   createdAt: currentMessage.createdAt, // <- NEW, to retain correct type of `createdAt`
>   revisionId: generateId2()
> };
> ```

Or, you could use `structuredClone` as originally implemented

**Comment by lgrammel** [comment]
@tobiasstrebitzer thanks, good find! https://github.com/vercel/ai/pull/3964
